pricegraph/src/orderbook.rs

```diff
@@ -129,15 +130,80 @@ impl Orderbook {
 
     /// Reduces the orderbook by matching all overlapping ring trades.
     pub fn reduce_overlapping_orders(&mut self) {
+        Subgraphs::new(self.projection.node_indices())
+            .for_each(|token| self.reduced_shortest_paths(token));
+    }
+
+    /// Fill a market order in the current orderbook graph returning the maximum
+    /// price the order can have while overlapping with existing orders. Returns
+    /// `None` if the order cannot be filled because the token pair is not
+    /// connected, or the connection cannot support enough capacity to fully
+    /// fill the specified volume.
+    ///
+    /// This is done by pushing the specified volume as a flow through the graph
+    /// and finding the min-cost paths that have enough capacity to carry the
+    /// flow from the sell token to the buy token. The min-cost is calculated
+    /// using the successive shortest path algorithm with Bellman-Ford.
+    ///
+    /// Note that this method is not idempotent, successive calls with the same
+    /// token pair and volume yield different results as orders get filled with
+    /// each match changing the orderbook.
+    pub fn fill_market_order(&mut self, pair: TokenPair, volume: f64) -> Option<f64> {
         self.update_projection_graph();
 
-        Subgraphs::new(self.projection.node_indices()).for_each(|token| loop {
-            let mut negative_cycle_predecessor = None;
-            match bellman_ford::search(&self.projection, token) {
-                Ok((_, predecessor)) => break negative_cycle_predecessor.unwrap_or(predecessor),
-                Err(NegativeCycle(predecessor, start)) => {
+        let (sell, buy) = (node_index(pair.sell), node_index(pair.buy));
+        let predecessors = self.reduced_shortest_paths(sell);
+        let mut path = path::find_path(&predecessors, sell, buy)?;
+
+        if volume == 0.0 {
```

Comment:
Should this be the first thing the function does as none of the stuff above is necessary?

Reply:
I mentioned this in the comment that I do the stuff above first so that we return `None` when there is no path, regardless of volume. I'm OK either way.

I had also considered having a volume of 0 be used so that you get the lowest transitive price between two orders without modifying the orderbook.

Let me know what you think is better.

Reply:
Oh I see now. I missed the ? which actually returns something before this code is hit. I think I'd slightly prefer to get the lowest possible transitive price since for technically for 0 volume any price is correct (the buy amount would always be 0 = (sellAmount = 0)* price). 

Maybe the notion of sending the smallest possible epsilon of flow through the network makes more sense?